ic-canister/ic-canister-macros/src/derive.rs

```diff
@@ -193,36 +193,48 @@ fn expand_upgrade_methods(
     };
 
     let (state_get, state_borrow) = (
-        quote! { let #name = #field_type::get(); },
+        quote! { let #name = ::std::rc::Rc::clone(&self. #name); },
         quote! { &* #name.borrow(), },
     );
 
-    let field_assignment = quote! { #field_type::get().replace(#name); };
+    let field_assignment = quote! { self. #name.replace(#name); };
 
     quote! {
         impl #struct_name {
-            #[cfg(all(target_arch = "wasm32", feature = "export_api"))]
+            #[cfg(feature = "export_api")]
             #[export_name = "canister_pre_upgrade"]
             fn __pre_upgrade() {
+                let instance = Self::init_instance();
+                instance.__pre_upgrade_inst();
+            }
+
+            fn __pre_upgrade_inst(&self) {
```

Comment:
One thing we could consider doing is to add these methods to the `Canister` trait, so that we don't have to have magic `__pre_upgrade_inst` in the tests.

In case there's no stable state, we can just generate empty methods (as oposed to skipping them altogether in current implementation)